### PR TITLE
Convert pullSecrets to namespacedNames

### PR DESCRIPTION
**Pull request by psav**

### CODE REVIEW

File: controllers/cloud.redhat.com/clowdapp_controller.go [hunk 1]
```diff
@@ -427,21 +428,25 @@ func contains(list []string, s string) bool {
 	return false
 }
 
-func (r *ClowdAppReconciler) runProviders(provider *providers.Provider, a *crd.ClowdApp) error {
+func (r *ClowdAppReconciler) runProviders(log logr.Logger, provider *providers.Provider, a *crd.ClowdApp) error {
```
Comment (BlakeHolifield) [inline controllers/cloud.redhat.com/clowdapp_controller.go:431]: It seems like you already get this from the struct on line 77. What's the advantage of passing this in vs using `r.Log....`?

File: controllers/cloud.redhat.com/clowdapp_controller.go [hunk 2]
```diff
@@ -427,21 +428,25 @@ func contains(list []string, s string) bool {
...
 
 	c := config.AppConfig{}
 
 	for _, provAcc := range providers.ProvidersRegistration.Registry {
+		// log.Info("running provider: fart")
```
Comment (BlakeHolifield) [inline controllers/cloud.redhat.com/clowdapp_controller.go:436]: lolz

File: controllers/cloud.redhat.com/clowdenvironment_controller.go [hunk 3]
```diff
@@ -218,8 +219,9 @@ func (r *ClowdEnvironmentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, e
 	return ctrl.Result{Requeue: requeue}, nil
 }
 
-func runProvidersForEnv(provider providers.Provider) error {
+func runProvidersForEnv(log logr.Logger, provider providers.Provider) error {
```
Comment (BlakeHolifield) [inline controllers/cloud.redhat.com/clowdenvironment_controller.go:222]: Same `r.Log` vs `Env(log)` question.

File: controllers/cloud.redhat.com/providers/pullsecrets/provider.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+// CoreEnvPullSecrets is the pull_secrets for the app.
+var CoreEnvPullSecrets = providers.NewMultiResourceIdent(ProvName, "core_env_pull_secrets", &core.Secret{})
+
+// GetPullSecret returns the correct end provider.
+func GetPullSecret(c *providers.Provider) (providers.ClowderProvider, error) {
+	return NewPullSecretProvider(c)
+}
+
+func init() {
+	providers.ProvidersRegistration.Register(GetPullSecret, 98, ProvName)
```
Comment (BlakeHolifield) [inline controllers/cloud.redhat.com/providers/pullsecrets/provider.go:20]: Should we consider refactoring this out to an enum or some layer/tier system? I know 98 is later in the process, but I don't know how much later it is exactly.

File: controllers/cloud.redhat.com/providers/serviceaccount/default.go [hunk 5]
```diff
@@ -88,7 +88,7 @@ func (sa *serviceaccountProvider) Provide(app *crd.ClowdApp, c *config.AppConfig
 
 		labeler := utils.GetCustomLabeler(nil, nn, app)
 
-		if err := CreateServiceAccount(sa.Cache, CoreDeploymentServiceAccount, sa.Env.Spec.Providers.PullSecrets, nn, labeler); err != nil {
+		if err := CreateServiceAccount(sa.Cache, CoreDeploymentServiceAccount, nn, labeler); err != nil {
```
Comment (BlakeHolifield) [inline controllers/cloud.redhat.com/providers/serviceaccount/default.go:91]: We're gonna have conflicts here depending on who gets merged first. Just a heads up

File: controllers/cloud.redhat.com/providers/pullsecrets/default.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+		newPullSecObj := &core.Secret{}
+
+		secName := fmt.Sprintf("%s-clowder-copy", pullSecretName.Name)
+
+		newSecNN := types.NamespacedName{
+			Name:      secName,
+			Namespace: targetNamespace,
+		}
+
+		secList = append(secList, secName)
```
Comment (BlakeHolifield) [inline controllers/cloud.redhat.com/providers/pullsecrets/default.go:136]: Do we need to append the `clowder-copy` secret if we aren't going to create it?

**Review by BlakeHolifield — CHANGES_REQUESTED** [review]

File: controllers/cloud.redhat.com/clowdapp_controller.go (hunk 1 above)
Comment (psav) [reply controllers/cloud.redhat.com/clowdapp_controller.go:431]: log context

File: controllers/cloud.redhat.com/providers/pullsecrets/provider.go (hunk 4 above)
Comment (psav) [reply controllers/cloud.redhat.com/providers/pullsecrets/provider.go:20]: We _could_ but then how would we do half tiers.....this is a pretty delicate piece of the system

File: controllers/cloud.redhat.com/providers/pullsecrets/default.go (hunk 6 above)
Comment (psav) [reply controllers/cloud.redhat.com/providers/pullsecrets/default.go:136]: yes - this is the list of secrets that get added to the sa

File: controllers/cloud.redhat.com/providers/serviceaccount/default.go (hunk 5 above)
Comment (psav) [reply controllers/cloud.redhat.com/providers/serviceaccount/default.go:91]: Yeh

File: controllers/cloud.redhat.com/providers/pullsecrets/default.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+// NewPullSecretProvider returns a new End provider run at the end of the provider set.
+func NewPullSecretProvider(p *providers.Provider) (providers.ClowderProvider, error) {
+
+	appList := &crd.ClowdAppList{}
+	if err := p.Env.GetAppsInEnv(p.Ctx, p.Client, appList); err != nil {
+		return nil, err
+	}
+
+	namespaceList := map[string]bool{}
```
Comment (kylape) [inline controllers/cloud.redhat.com/providers/pullsecrets/default.go:31]: maybe rename to namespaceSet?

File: controllers/cloud.redhat.com/providers/pullsecrets/default.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+		for namespace := range namespaceList {
+			if namespace == pullSecretName.Namespace {
+				secList = append(secList, pullSecretName.Name)
+				continue
+			}
+
+			sourcePullSecObj := &core.Secret{}
+			if err := prov.Client.Get(prov.Ctx, types.NamespacedName{
+				Name:      pullSecretName.Name,
+				Namespace: pullSecretName.Namespace,
```
Comment (kylape) [inline controllers/cloud.redhat.com/providers/pullsecrets/default.go:70]: why not do this in the out for loop?

File: controllers/cloud.redhat.com/providers/pullsecrets/default.go [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+			}, sourcePullSecObj); err != nil {
+				return nil, err
+			}
+
+			newPullSecObj := &core.Secret{}
+
+			secName := fmt.Sprintf("%s-clowder-copy", pullSecretName.Name)
```
Comment (kylape) [inline controllers/cloud.redhat.com/providers/pullsecrets/default.go:77]: maybe add env in the name?

File: controllers/cloud.redhat.com/providers/pullsecrets/default.go [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+
+	return &pullsecretProvider{Provider: *p}, nil
+}
+
+func copyPullSecrets(prov *providers.Provider, namespaceList map[string]bool) ([]string, error) {
+	var secList []string
+
+	for _, pullSecretName := range prov.Env.Spec.Providers.PullSecrets {
+		for namespace := range namespaceList {
+			if namespace == pullSecretName.Namespace {
```
Comment (kylape) [inline controllers/cloud.redhat.com/providers/pullsecrets/default.go:62]: let's not make an exception for this.  just copy it to a new secret in the same namespace if applicable

**Review by kylape — CHANGES_REQUESTED** [review]

**Pull request merged**